recorder.go

```diff
@@ -588,19 +644,28 @@ func (r *Recorder) shouldFlush() bool {
 
 func (r *Recorder) reportLoop() {
 	tickerChan := time.Tick(minReportingPeriod)
-	for range tickerChan {
-		r.maybeLogInfof("reporting alarm fired")
-
-		// Kill the reportLoop() if we've been disabled.
-		r.lock.Lock()
-		if r.disabled {
+	for {
+		select {
+		case <-tickerChan:
+			now := time.Now()
+
+			r.lock.Lock()
+			disabled := r.disabled
+			reconnect := !r.reportInFlight && now.Sub(r.connTimestamp) > r.reconnectPeriod
+			shouldFlush := r.shouldFlushLocked(now)
 			r.lock.Unlock()
-			break
-		}
-		r.lock.Unlock()
 
-		if r.shouldFlush() {
-			r.Flush()
+			if disabled {
+				return
+			}
+			if reconnect {
+				r.reconnectClient(now)
+			}
+			if shouldFlush {
+				r.Flush()
```

Comment:
not sure if this would actually be an issue, but would it make sense to flush and then reconnectClient? That way  in case the reconnect doesn't work at least you have flush what you currently have?